fiftyone/utils/aws.py

```diff
@@ -0,0 +1,127 @@
+"""
+Utilities for working with
+`Amazon Web Services <https://aws.amazon.com/>`
+S3 storage buckets.
+
+| Copyright 2017-2021, Voxel51, Inc.
+| `voxel51.com <https://voxel51.com/>`_
+|
+"""
+import logging
+import multiprocessing
+import multiprocessing.dummy
+import os
+from urllib.parse import urlparse
+
+import boto3
+import botocore
+
+import fiftyone.core.utils as fou
+
+import eta.core.utils as etau
+
+
+logger = logging.getLogger(__name__)
+
+
+def download_from_s3(
+    urls, download_dir=None, num_workers=None,
+):
+    """Download files from a public AWS S3 bucket using unsigned urls.
+
+    The `url` argument either accepts:
+
+        * A list of paths to objects in the s3 bucket::
+
+            urls = ["s3://bucket_name/dir1/file1.ext", ...]
+
+          When `urls` is a list, then the `download_dir` argument is required
+          and all objects will be downloaded into that directory
+
+        * A dictionary mapping the paths of objects to files on disk
+          to store each object::
+
+            urls = {
+                "s3://bucket_name/dir1/file1.ext": "/path/to/local/file1.ext",
+                ...
+            }
+
+    Args:
+        urls: either a list of urls to objects in an s3 bucket, or a dict
+            mapping these urls to locations on disk. If `urls` is a list, then
+            the `download_dir` argument is required
+        download_dir (None): the directory to store all downloaded objects.
+            This is only used if `urls` is a list
+        num_workers (None): the number of processes to use when downloading
+            files. By default, ``multiprocessing.cpu_count()`` is used
+    """
+    if isinstance(urls, list):
+        if download_dir is None:
+            raise ValueError(
+                "When `urls` is a list, `download_dir` is required but was found to be `None`."
+            )
+        urls = {url: None for url in urls}
+
+    if download_dir:
+        etau.ensure_dir(download_dir)
+
+    if num_workers is None or num_workers < 1:
+        num_workers = multiprocessing.cpu_count()
+
+    s3_client = boto3.client(
+        "s3",
+        config=botocore.config.Config(
+            signature_version=botocore.UNSIGNED,
+            max_pool_connections=max(10, num_workers),
+        ),
+    )
+
+    inputs = _build_inputs(urls, s3_client, download_dir=download_dir)
+    num_samples = len(inputs)
+
+    if num_samples == 0:
+        return 0
+
+    if num_workers == 1:
+        _single_thread_download(inputs)
+    else:
+        _multi_thread_download(inputs, num_workers)
+
+    return num_samples
```

Comment:
The way this function is written, the user will always know exactly how many files will be downloaded (also this return value isn't documented), so I'd just omit. No return value needed